src/core/instance/render-helpers/render-list.js

```diff
@@ -9,23 +9,29 @@ export function renderList (
   val: any,
   render: () => VNode
 ): ?Array<VNode> {
-  let ret: ?Array<VNode>, i, l, keys, key
+  let ret: ?Array<VNode>, i, l, keys, key, vnode
   if (Array.isArray(val) || typeof val === 'string') {
     ret = new Array(val.length)
     for (i = 0, l = val.length; i < l; i++) {
-      ret[i] = render(val[i], i)
+      vnode = render(val[i], i)
+      vnode.isListItem = true
+      ret[i] = vnode
     }
   } else if (typeof val === 'number') {
     ret = new Array(val)
     for (i = 0; i < val; i++) {
-      ret[i] = render(i + 1, i)
+      vnode = render(i + 1, i)
+      vnode.isListItem = true
+      ret[i] = vnode
     }
   } else if (isObject(val)) {
     keys = Object.keys(val)
     ret = new Array(keys.length)
     for (i = 0, l = keys.length; i < l; i++) {
       key = keys[i]
-      ret[i] = render(val[key], key, i)
+      vnode = render(val[key], key, i)
+      vnode.isListItem = true
+      ret[i] = vnode
     }
   }
```

Comment:
Instead of adding a property to every vnode, we can simply mark the `ret` array here with something like:

``` js
if (isDef(ret)) {
  (ret: any)._isVList = true // work around flow
}
```

Then check `children._isVList` in `normalizeArrayChildren`.